utils_cv/action_recognition/dataset.py

```diff
@@ -0,0 +1,473 @@
+        """
+        Args:
+            record (VideoRecord): A video record.
+
+        Return:
+            list: Segment offsets (start indices)
+        """
+        if record.num_frames > self.presample_length:
+            if self.random_shift:
+                # Random sample
+                offsets = np.sort(
+                    randint(
+                        record.num_frames - self.presample_length + 1,
+                        size=self.num_segments,
+                    )
+                )
+            else:
+                # Uniform sample
+                distance = (
+                    record.num_frames - self.presample_length + 1
+                ) / self.num_segments
+                offsets = np.array(
+                    [
+                        int(distance / 2.0 + distance * x)
+                        for x in range(self.num_segments)
+                    ]
+                )
+        else:
+            if self.warning:
+                warnings.warn(
+                    "num_segments and/or sample_length > num_frames in {}".format(
+                        record.path
+                    )
+                )
+            offsets = np.zeros((self.num_segments,), dtype=int)
+
+        return offsets
+
+    def _get_frames(
+        self, video_reader: decord.VideoReader, offset: int,
+    ) -> List[np.ndarray]:
+        """ Get frames at sample length.
+
+        Args:
+            video_reader: the decord tool for parsing videos
+            offset: where to start the reader from
+
+        Returns
+            Frames at sample length in a List
+        """
+        clip = list()
+
+        # decord.seek() seems to have a bug. use seek_accurate().
+        video_reader.seek_accurate(offset)
+        # first frame
+        clip.append(video_reader.next().asnumpy())
+        # remaining frames
+        try:
+            if self.temporal_jitter:
+                for i in range(self.sample_length - 1):
+                    step = randint(self.sample_step + 1)
+                    if step == 0:
+                        clip.append(clip[-1].copy())
+                    else:
+                        if step > 1:
+                            video_reader.skip_frames(step - 1)
+                        cur_frame = video_reader.next().asnumpy()
+                        if len(cur_frame.shape) != 3:
+                            # maybe end of the video
+                            break
+                        clip.append(cur_frame)
+            else:
+                for i in range(self.sample_length - 1):
+                    if self.sample_step > 1:
+                        video_reader.skip_frames(self.sample_step - 1)
+                    cur_frame = video_reader.next().asnumpy()
+                    if len(cur_frame.shape) != 3:
+                        # maybe end of the video
+                        break
+                    clip.append(cur_frame)
+        except StopIteration:
```

Comment:
[minor] add comment here when this exception is thrown (maybe once the end of the video is reached?)